[Samuel Tee] iP

Kidsnd274 opened the pull request:
# Duke Task List Assistant 
Duke Task List Assistant helps you keep track of your tasks and more!

## Description
An advanced task list assistant that is flexible and expandable. (Runs in Command Line and GUI modes)

```
Hello from
 ____        _        
|  _ \ _   _| | _____ 
| | | | | | | |/ / _ \
| |_| | |_| |   <  __/
|____/ \__,_|_|\_\___|

<---------------------------------------------------------->
Hello! I'm Duke, your personal assistant
What can I do for you?
<---------------------------------------------------------->
```

### Features 👓
Currently Supported Types of Tasks:
- Todo Task (normal task with a description)
- Event Task (event task with a date)
- Deadline Task (deadline task with a deadline)
- Supports exporting and importing the task list with a file
- GUI Support Coming Soon! 😊

### Downloads 🔽
- [Releases](https://github.com/Kidsnd274/ip/releases)

### Usage ⌨
Command List: _(for CMD mode)_ 📜
- list
- delete (task number)
- todo (description)
- event (description) /at (date in YYYY-MM-DD format)
- deadline (description) /by (date in YYYY-MM-DD format)
- mark (task number)
- unmark (task number)
- find (keyword)
- exit

Some example commands:
```
list
todo borrow book
deadline return book /by 1999-03-03
event project meeting at /at 2022-01-24
list
mark 2
mark 3
list
mark 1
unmark 2
list
quit
```

GUI Mode:
> COMING SOON! 😊

Exporting and importing tasklists:
1. Duke saves to task.txt for every modification to the task list
2. Duke will immediately import tasks from task.txt on startup

### Some code snippets: 👩‍💻
Startup run code:
```java
 public void run() {
     // Setting up the goods
     ui = new TextUi();
     fh = new Storage(FILENAME);
     taskList = new TaskList(fh.importTasks());
     parser = new Parser();
     parser.setTaskList(taskList);

     ui.printIntro();

     // Starting input loop
     while (true) {
         String input = ui.getInput();
         if (processInput(input)) {
             break;
         }
     }
     ui.printGoodbye();
 }
 ```
~~Hi~~ Bye **(Honestly don't know what I'm supposed to add to bold, italic and strikethrough)**

Commit:
Add Gradle support

Kidsnd274 committed:
- Edited .gitignore to exclude Java .class files
- Implemented Duke Level-1

Kidsnd274 committed:
- Implemented Level-2: Add, List

Kidsnd274 committed:
- Implemented Level-3: Mark as Done
- Added Task Class
- Added new checks for mark and unmark
ISSUES:
- Many invalid inputs can cause a crash

Kidsnd274 committed:
Implemented Level-4. ToDos, Events, Deadlines
- Added new classes Todo, Events and Deadlines
- Moved input processing to separate method
- Added separate method for generating list
- Moved some text as static variables
- Task has more methods

Kidsnd274 committed:
Added A-TextUiTesting

Kidsnd274 committed:
Level-5: Handle Errors
- Reorganized code into packages
- Created custom exceptions
- Edited runtest.bat to pause
- Added new method to check for invalid number inputs

Kidsnd274 committed:
Level 6. Delete
- Fixed a bug where if no index is entered, program will crash
- Added delete method and generate message for delete

Todo:
- Haven't written test case for delete and error checking

Kidsnd274 committed:
Added Enum
- Added InputEnum but not used yet

Kidsnd274 committed:
Move and reorganize code into packages

Made some variables private

Kidsnd274 committed:
Add functionality to save tasklist to a file
Tasklist is not persistent and is cleared everytime the program closes.

With this new change, the tasklist will now be saved and the list will persist through multiple runs

Reads from the file on launch and saves for every change to the tasklist

Changes:
- Add new functions and constructors for Tasks to support adding existing tasks from file
- New saving and reading functionality (tasklist now saves)
- New FileHandler class which deals with reading and writing the tasklist file
- Add changes to runtest.bat to include testing for the tasklist file
- Add task.txt to .gitignore

Kidsnd274 committed:
Merge branch 'branch-Level-7'

Kidsnd274 committed:
Replace string representation of date with proper date object
Date was represented by string.

With this new change, a date object will be used to store the date instead.

Changes:
- Add new abstract TaskWithDate object to further OOP
- Add new getDate method for Tasks with Dates

Kidsnd274 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/kidsnd274/duke/taskobjects/Deadline.java
#	src/main/java/kidsnd274/duke/taskobjects/Task.java
#	text-ui-test/runtest.bat

Kidsnd274 committed:
Clean up merge with Level 7 and Level 8
There were some code issues between both commits

Kidsnd274 committed:
Implemented TextUi class
The main Duke class used to handle all the input and output

New class has been made to abstract the UI functions from the main class

Changes:
- Created new classes for Parser, TaskList, TextUi
- Changed FileHandler to Storage (name)

Kidsnd274 committed:
Restructure entire program to make it more OOP
Main class Duke used to handle all the main logic. Now it's split to its individual categories.

This makes expansion for each element harder in the future

To make the code more OOP, each function has it's own class and tree structure. UI, Storage, Parser, TaskList,

Changes:
- Coded Parser
- Added TextUi class which handles all input and output
- TaskList abstraction added
- Command package added. Each command has it's own class
- Removed unnecessary InputEnum.java

To Do:
- Add more functionality to CommandResult and get rid of getTaskList function from TaskList (encapsulation issues)

Kidsnd274 committed:
Move TextUi to a separate package (to allow for other UIs in the future)

Kidsnd274 committed:
Add JUnit Testing and reorganized some modules

Kidsnd274 committed:
Add support for JAR building

Kidsnd274 committed:
Add JavaDocs for most classes except for classes in the command package

Kidsnd274 committed:
Add more javadocs and fixed some coding standards

Kidsnd274 committed:
Add search/find functionality

Duke could not search for tasks

Now Duke can use a key word to search for tasks with that keyword

Changes:
- Added FindCommand class
- Added search functionality
- Updated test case to test for search

Kidsnd274 committed:
Merge branch 'branch-A-JavaDoc'

Kidsnd274 committed:
Merge branch 'branch-A-CodingStandard'

Kidsnd274 committed:
Merge branch 'branch-Level-9'

Kidsnd274 committed:
Update README.md

Kidsnd274 committed:
Fix readme

Addition W

Kidsnd274 committed:
Add Gradle and build functionality with shadowJar
IntellJ used to handle all the building and running test cases

Using Gradle will improve future workflow and less manual running of test cases and building. It will also handle all the dependencies easily

Gradle is now used for building and running JUnit tests

Changes:
- Changed main package from kidsnd274.duke to duke
- Added gradle files and gradle config
-- Use gradle build, gradle clean, gradle shadowJar

Todo:
- Use gradle to run test.bat to check CMD output
- Transition tests to JUnit

Kidsnd274 committed:
Fix gradle run

Kidsnd274 committed:
Merge branch 'branch-add-gradle-support'

cheehongw commented on src/main/java/kidsnd274/duke/Parser.java:
```diff
@@ -0,0 +1,160 @@
+     *
+     * It also checks for erroneous input
+     * @param input User input
+     * @return A {@code Command} type based on user's input
```
should have punctuation behind each parameter description.

cheehongw commented on src/main/java/kidsnd274/duke/Storage.java:
```diff
@@ -0,0 +1,108 @@
+package kidsnd274.duke;
+
+import kidsnd274.duke.taskobjects.*;
```
don't use wildcard imports

Kidsnd274 committed:
Add JUnit Test Cases for Parser
StorageTest.java also added but not implemented yet
Made some changes to build.gradle as well

Kidsnd274 committed:
Merge branch 'add-gradle-support' of https://github.com/Kidsnd274/ip

# Conflicts:
#	build.gradle
#	gradle/wrapper/gradle-wrapper.jar
#	gradle/wrapper/gradle-wrapper.properties
#	gradlew
#	gradlew.bat

Kidsnd274 committed:
Fix ParserTest's JUnit tests

cheehongw commented on src/main/java/kidsnd274/duke/TaskList.java:
```diff
@@ -0,0 +1,115 @@
+    /**
+     * Adds a new {@code Task} to the task list
+     * @param task Provided Task that is added to the task list
+     * @return Returns true is addition is successful, else false
```
should have a blank line between description and parameter section.

The javadocs for all your methods in `TaskList.java` seems to have this problem.

Kidsnd274 committed:
Update runtest.bat with new directory mappings

cheehongw commented on src/main/java/kidsnd274/duke/Storage.java:
```diff
@@ -0,0 +1,108 @@
+                newTask = new Event(description, isDone, inputArray[3]);
+            } else if (type.equals("D")) {
+                newTask = new Deadline(description, isDone, inputArray[3]);
+            }
```
maybe a switch-case statement here would look neater? 🤔

Kidsnd274 committed:
Add Checkstyle functionality

Kidsnd274 committed:
Merge branch 'branch-checkstyle'

# Conflicts:
#	build.gradle

cheehongw reviewed (commented):
LGTM 😄 

I can identify some javadocs don't meet the coding standards of this module. It would be good to setup checkstyle to catch these kinds of coding standard violation.

Kidsnd274 committed:
Add JavaFx and implement both GUI and CMD modes
GUI mode incomplete

Kidsnd274 committed:
Fix some checkstyle issues

Kidsnd274 committed:
Implement new GUI features (not complete yet)
GUI does not have any features

Main backbone framework of GUI has been implemented

Functionality will be added soon

ckcherry23 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,69 @@
+        // Starting input loop
+        while (true) {
+            String input = ui.getInput();
+            if (processInput(input)) {
```
Since `processInput()` returns a `Boolean` value, perhaps you can use another function with a name like `isInputExitCommand()` and separate the logic for processing the input for other commands?

ckcherry23 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,69 @@
+        }
+        return false;
+    }
+}
```
Can end your files with a newline, as suggested by CheckStyle.

ckcherry23 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,170 @@
+import duke.command.MarkAsUndoneCommand;
+import duke.exceptions.NullDateProvidedException;
+
+import java.time.LocalDate;
```
_Suggestion:_ I like how you have grouped your imports! But in case you are working for team projects, following the CheckStyle ordering for `import` statements would help you keep the ordering uniform compared to your teammates

ckcherry23 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,113 @@
+package duke;
+
+import duke.taskobjects.*;
```
Do we need to import everything in `duke.taskobjects`, or is this a wildcard import?

ckcherry23 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,115 @@
+     * Returns the contained task list. Terrible function (will remove soon)
+     * @return The contained task list
+     */
+    public List<Task> getList() { // GET RID OF THIS IN THE FUTURE
```
Perhaps you can mark such personal comments with `//TODO` or `//FIXME` and view them in the **TODO tool window** in **IntelliJ**?

ckcherry23 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,115 @@
+        return newString.toString();
+    }
+
+//    public boolean markAsDone(int taskNo) {
+//        Task currentTask = taskList.get(taskNo);
+//        currentTask.markAsDone();
+//        return true;
+//    }
+//
+//    public boolean markAsUndone(int taskNo) {
+//        Task currentTask = taskList.get(taskNo);
+//        currentTask.markAsUndone();
+//        return true;
+//    }
+}
```
I think you can remove the code commented out as it has been implemented elsewhere.

ckcherry23 reviewed (commented):
Looks good! 👍🏼
Just a few suggestions and some nits to fix.

Kidsnd274 committed:
Implement GUI features but buttons are not programmed
Converted to FXML

Issues:
- ImageBox in DialogBox doesn't have a clean circle cut
- Loading FXML document with JavaFX API of version 17 by JavaFX runtime of version 11

Kidsnd274 committed:
Connect GUI to Duke properly
GUI only responded with user's input

Does not actually work as a task list

With the new change, the GUI can now work as a task list

Changes:
- Remove unnecessary Main.java file
- Make some UI changes to allow text wrapping
- Some changes to Duke

Kidsnd274 committed:
Merge branch 'branch-gui'

# Conflicts:
#	build.gradle

Kidsnd274 committed:
Fix some checkstyle issues

Add CommandTest class as well (not implemented yet)

Kidsnd274 committed:
Fix some JavaDocs formatting issues

Kidsnd274 committed:
Fix more JavaDocs stuff

Kidsnd274 committed:
Fix some issues with build.gradle

Kidsnd274 committed:
Add assertions in various points of the code

There are several methods with calculations of an index of an array. There are also cases where a method is assumed to mark an task as done or undone. These calculations and methods are assumed to work correctly.

Using assertions in these methods ensure that these calculations of indexes are within range and that tasks are marked correctly as well.

Assertions are used after the calculations or method call.

build.gradle updated to enable Assertions in testing and running

Kidsnd274 committed:
Merge pull request #1 from Kidsnd274/branch-A-Assertions

Add assertions in various points of the code

Kidsnd274 committed:
Refactor code from Parser.java

The parseCommand method is considered a long method with very hard to follow logic.

Splitting the method to smaller methods would improve code quality.

By changing the if-else statement to a switch-case and moving most of the logic to other smaller methods, code quality of parseCommand would improve. parseCommand would also have the same level of abstraction.

Also clean up some code where there are unnecessary else statements and merged some try-catch blocks as well.

Kidsnd274 committed:
Refactor code from Storage.java

Multiple methods were long and convoluted.

Splitting the method to smaller methods would improve code quality.

By changing the if-else statement to a switch-case and moving most of the logic to other smaller methods, code quality would improve.

Also created a ImportTaskParameters class to store parameter details for an imported task.
Also fixed an issue where the task list wasn't saving in GUI mode.
Also fixed an FXML error

Kidsnd274 committed:
Merge branch 'master' into branch-A-CodeQuality

Kidsnd274 committed:
Merge pull request #2 from Kidsnd274/branch-A-CodeQuality

Merging Code Quality Branch

Kidsnd274 committed:
Add Postpone feature

In previous versions, editing the dates for Event and Postpone tasks is not possible. The only way to do it is by deleting the task and adding it back with the correct date.

With this new version, the "postpone" keyword can be used to edit the date for those types of tasks. The command syntax is "postpone TASKNO DATE"

Kidsnd274 committed:
Add more test cases to ParserTest for Postpone function

Kidsnd274 committed:
Merge pull request #3 from Kidsnd274/branch-extension-b-snooze

Branch extension b snooze

Kidsnd274 committed:
Improve GUI
Add new user icons, add welcome message, reorganized some code, made changes to the FXML code with padding and chat bubbles, reduce size of profile icon

Fix a bug in DeleteCommand. It used to print the wrong number of items in the list left.

Kidsnd274 committed:
Add Error Message colours
DialogBox will now turn Orange when an error message is displayed.
Submitting blank will not submit a command

Kidsnd274 committed:
Merge branch 'branch-bettergui'

Kidsnd274 committed:
Fix JavaDocs issues

Kidsnd274 committed:
Fix issue where user profile image gets clipped by scrollbar

Kidsnd274 committed:
Add screenshot and edit repo README

Kidsnd274 committed:
Set theme jekyll-theme-cayman

Kidsnd274 committed:
Add new app icon. Update User guide and repo README

Kidsnd274 committed:
Merge branch 'master' of https://github.com/Kidsnd274/ip

Kidsnd274 committed:
Comment out some code

Kidsnd274 committed:
Add screenshots to user guide. Increase JavaFX version from 11 to 14 to fix font issues on Mac
